vnext/Microsoft.ReactNative/Utils/TextHitTestUtils.cpp

```diff
@@ -0,0 +1,328 @@
+        L = m + 1;
+      } else if (rect.X >= target.X) {
+        // If median character is right of end, boundary is before median
+        R = m - 1;
+      } else {
+        // Otherwise, we've found the boundary where the next character is the
+        // first RTL character and the median is the last LTR character
+        BidirectionalTextBoundary result;
+        result.rtlPointer = nextPointer;
+        result.rtlRect = nextRect;
+        result.ltrPointer = textPointer;
+        result.ltrRect = rect;
+        return result;
+      }
+    } else {
+      break;
+    }
+  }
+
+  // No LTR characters on last line, L points to first character on last line
+  auto lastRTL = end;
+  auto lastRTLRect = target;
+  if (L < end.Offset()) {
+    const auto relativeOffset = L - end.Offset();
+    lastRTL = end.GetPositionAtOffset(relativeOffset, xaml::Documents::LogicalDirection::Forward);
+    lastRTLRect = lastRTL.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+  }
+
+  BidirectionalTextBoundary result;
+  result.rtlPointer = lastRTL;
+  result.rtlRect = lastRTLRect;
+  return result;
+}
+
+/**
+ * Determine if the next character in the given direction on the same line is
+ * RTL. If the next character is not on the same line, assume LTR. This may be
+ * The case if the run has only one character on the target line.
+ */
+static bool IsRTL(const xaml::Documents::TextPointer &textPointer, const winrt::Rect &rect, int direction) {
+  auto currentPointer = textPointer.GetPositionAtOffset(direction, xaml::Documents::LogicalDirection::Forward);
+  while (currentPointer != nullptr) {
+    const auto currentRect = currentPointer.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+    if (currentRect.Y != rect.Y) {
+      return false;
+    } else if (currentRect.X != rect.X) {
+      return direction > 0 ? currentRect.X < rect.X : currentRect.X > rect.X;
+    }
+
+    currentPointer = currentPointer.GetPositionAtOffset(direction, xaml::Documents::LogicalDirection::Forward);
+  }
+
+  return false;
+}
+
+/**
+ * Determine if the character is either the first or last character of a line in the given direction.
+ */
+static std::optional<winrt::Rect>
+GetBoundaryCharacter(const xaml::Documents::TextPointer &textPointer, const winrt::Rect &rect, int direction) {
+  const auto nextPosition = textPointer.GetPositionAtOffset(direction, xaml::Documents::LogicalDirection::Forward);
+  if (nextPosition) {
+    const auto nextRect = nextPosition.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+    if (nextRect.Y == rect.Y) {
+      return nextRect;
+    }
+  }
+
+  return std::nullopt;
+}
+
+bool IsInBounds(const winrt::Point &point, float left, float right, float top, float bottom) {
+  return left <= point.X && right >= point.X && top <= point.Y && bottom >= point.Y;
+}
+
+bool TextHitTestUtils::HitTest(const xaml::Documents::Run &run, const winrt::Point &targetPoint) {
+  const auto start = run.ContentStart();
```

Comment:
This file includes a lot of algorithmia (some pre-existing, some new). Text is hard, as I'm sure we all know by know :)So the amount of code and complexity required to get it right is... a lot, and it becomes hard to follow. 

I expect this will be hard to maintain without tests.

So, I would love to have some basic automated tests (maybe even without rendering pixels on screen, can we do just a data model and mock the ui just so that we test the algorithm gives the right answers?) for the different cases the algorithm is accounting for, so that we don't regress it next time we need to go into this.

Reply:
Would a few integration tests be acceptable here? It looks like Microsoft.ReactNative is not set up for unit tests yet (happy to be wrong about this if you point me to the right project).

Reply:
Yes any kind of test would be good :) we have some native tests here:
- Microsoft.ReactNative.ComponentTests  
- Microsoft.ReactNative.Cxx.UnitTests   
- Microsoft.ReactNative.IntegrationTests

xaml doesn't usually lend itself too well to being unit-tested outside of a uwp test app but whatever you can come up with would be much appreciated!

Reply:
I added E2E tests as these felt like the lowest overall overhead w.r.t. setting up test infrastructure. The algorithm is based on TextPointer, and as you mentioned XAML doesn't lend itself well to unit testing. It felt like a lot of overhead to create abstractions and mock the text pointer, so I captured most of the edge cases in the E2E tests, including ones that we know do not work (they show up in the snapshot with `Pressed: 0 times` because the algorithm currently cannot hit test that point).